### ISSUE TITLE
Start providing a go-sdk

### ISSUE THREAD

**Issue by rmohr** [kubevirt/kubevirt#677]
In order to make it easier to integrate KubeVirt into existing go projects in the existing k8s ecosystem, it will make our life much easier, if we have a separate repo with client tools only, like it exists for k8s with client-go.

One example is the cluster autoscaler. To properly vendor our APIs and REST client into the autoscaler we need a trimmed down sdk which only depends on client-go and apimachinery from k8s. 

`pkg/kubecli` and `pkg/api` were already written with the idea of providing an sdk out of them in mind. Thy only have the above mentioned dependencies. Synchronizing them into another kubevirt repo should be very simple.

**Comment by rmohr** [comment]
@pkliczewski assigning that to you since I think you are looking into that.

**Comment by cynepco3hahue** [comment]
I think the issue is still relevant to the number of projects.
/remove-lifecycle stale

**Comment by kunalkushwaha** [comment]
Hi,  If no one is working on this, can I work on this?

**Comment by fabiand** [comment]
I think this work would be welcome.

**Comment by kunalkushwaha** [comment]
I started working on this.

Have moved the code (`pkg/api` & `pkg/kubecli`) in separate repo [go-sdk](https://github.com/kunalkushwaha/go-sdk) 

updated the reference of these packages in my branch https://github.com/kunalkushwaha/kubevirt/tree/go-sdk

The problem I am facing right now is bezel build error.

```
.
.
external/com_github_bazelbuild_buildtools: warning: directory does not exist.
Target //:buildifier up-to-date:
  bazel-bin/buildifier.bash
INFO: Elapsed time: 152.460s, Critical Path: 13.53s
INFO: 247 processes: 1 local, 246 processwrapper-sandbox.
INFO: Build completed successfully, 258 total actions
INFO: Build completed successfully, 258 total actions
ERROR: /root/go/src/kubevirt.io/kubevirt/cmd/subresource-access-test/BUILD.bazel:3:1: no such package '@com_github_kunalkushwaha_go_sdk//kubecli': The repository could not be resolved and referenced by '//cmd/subresource-access-test:go_default_library'
ERROR: Analysis of target '//cmd/subresource-access-test:subresource-access-test' failed; build aborted: no such package '@com_github_kunalkushwaha_go_sdk//kubecli': The repository could not be resolved
INFO: Elapsed time: 4.093s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (16 packages loaded, 3 targets configured)
    currently loading: @bazel_tools//tools/zip
    Fetching @libvirt; Restarting.
+ finish
+ docker stop b16e7803f98133a67da40b562c7a7473375ff1746aa54d1401d52ad8758f9e09
+ docker rm -f b16e7803f98133a67da40b562c7a7473375ff1746aa54d1401d52ad8758f9e09
Makefile:5: recipe for target 'all' failed
make: *** [all] Error 1

```

How can I resolve this? 
The packages are part of `vendor/github.com/..`, do I need to add somewhere in bezel files too?

**Comment by slintes** [comment]
Great that you're working on this :+1: 

Does your `go-sdk` compile? I don't see any dependency management like [glide](https://glide.sh/) configured, and no `vendor` directory. I think you need to set this up first, before you can use the sdk in kubevirt itself or other projects. Please use the same versions of dependencies like we do here, see `glide.yaml`.

When that's done, you need to add the sdk to kubevirt, also with `glide`: add it to `glide.yaml`, and run `make deps-update`. That will put the sdk into the kubevirt's `vendor` dir. After that bazel should be happy, too. 

If you need help for setting this up, let us know :slightly_smiling_face: 

/remove-lifecycle stale

**Comment by kunalkushwaha** [comment]
Thanks for pointers.

I think my yesterday issue was related old kubevirt code. Today, I updated the code and also added `vendor` go-sdk.

Somehow I messed up with vendors of kubevirt. Will try to resolve on Monday.

**Comment by kunalkushwaha** [comment]
I think, vendoring dependencies not working nicely as most data structure conflicts  with dual vendor e.g `kubevirt.io/kubevirt/vendor/k8s/api/..` with `kubevirt.io/kubevirt/vendor/github.com/kunalkushwaha/go-sdk/vendor/k8s.io/api/..`

After resolving all dependencies issues, I am getting this error.

```
hack/dockerized "DOCKER_PREFIX= DOCKER_TAG= IMAGE_PULL_POLICY= VERBOSITY= ./hack/build-manifests.sh && \
    hack/bazel-fmt.sh && hack/bazel-build.sh"
[+] Building 0.1s (10/10) FINISHED                                                                                                                                                             
 => [internal] load .dockerignore                                                                                                                                                         0.0s
 => => transferring context: 2B                                                                                                                                                           0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                      0.0s
 => => transferring dockerfile: 784B                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/kubevirt/builder@sha256:08b4db91e5f6272004892c6174129172c1bef5c6906bde9f8414cb945a588e12                                                       0.0s
 => [1/4] FROM docker.io/kubevirt/builder@sha256:08b4db91e5f6272004892c6174129172c1bef5c6906bde9f8414cb945a588e12                                                                         0.0s
 => [internal] load build context                                                                                                                                                         0.0s
 => => transferring context: 33B                                                                                                                                                          0.0s
 => [internal] helper image for file operations                                                                                                                                           0.0s
 => CACHED [2/4] RUN     mkdir -p /go &&     source /etc/profile.d/gimme.sh &&     go get github.com/mattn/goveralls &&     go get -u github.com/Masterminds/glide &&     go get -u gith  0.0s
 => CACHED [3/4] RUN pip install j2cli && pip3 install operator-courier==1.3.0                                                                                                            0.0s
 => CACHED [4/4] COPY rsyncd.conf /etc/rsyncd.conf                                                                                                                                        0.0s
 => exporting to image                                                                                                                                                                    0.0s
 => => exporting layers                                                                                                                                                                   0.0s
 => => writing image sha256:f100f1edf06fb56949e0b95fb5843d0a1ec616c1fe919993d8b4a99ce89de976                                                                                              0.0s
 => => naming to docker.io/library/kubevirt                                                                                                                                               0.0s
go version go1.11.5 linux/amd64

go version go1.11.5 linux/amd64
/root/go/src/kubevirt.io/kubevirt/tools/manifest-templator/manifest-templator flag redefined: log_dir
panic: /root/go/src/kubevirt.io/kubevirt/tools/manifest-templator/manifest-templator flag redefined: log_dir

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc000098180, 0x1508b20, 0xc000335940, 0x139928f, 0x7, 0x13c76a1, 0x2f)
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/flag/flag.go:805 +0x529
flag.(*FlagSet).StringVar(0xc000098180, 0xc000335940, 0x139928f, 0x7, 0x0, 0x0, 0x13c76a1, 0x2f)
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/flag/flag.go:708 +0x8a
flag.(*FlagSet).String(0xc000098180, 0x139928f, 0x7, 0x0, 0x0, 0x13c76a1, 0x2f, 0xc000335930)
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/flag/flag.go:721 +0x8b
flag.String(0x139928f, 0x7, 0x0, 0x0, 0x13c76a1, 0x2f, 0xc000739e01)
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/flag/flag.go:728 +0x69
Makefile:5: recipe for target 'all' failed
make: *** [all] Error 2

```

@slintes Any idea. 
I will create a Draft Pull request, so it can easily be reviewed.

**Comment by rmohr** [comment]
> I think, vendoring dependencies not working nicely as most data structure conflicts with dual vendor e.g `kubevirt.io/kubevirt/vendor/k8s/api/..` with `kubevirt.io/kubevirt/vendor/github.com/kunalkushwaha/go-sdk/vendor/k8s.io/api/..`

Yes, all vendoring managers support flattening dependencies. You can run `make deps-update` and `make deps-install` to update vendor dependencies. That calls `glide` the right way.

**Comment by rmohr** [comment]
> /root/go/src/kubevirt.io/kubevirt/tools/manifest-templator/manifest-templator flag redefined: log_dir
> panic: /root/go/src/kubevirt.io/kubevirt/tools/manifest-templator/manifest-templator flag redefined: log_dir

That still looks like you did not flatten the vendor dependencies. You probably have `glog` included in your vendor directory. Either use our `make` targets or run `hack/dep-prune.sh` manually.

**Comment by kunalkushwaha** [comment]
Thanks. I was finally able to resolve all vendor issues. 
Create a PR https://github.com/kubevirt/kubevirt/pull/2286

Please take a look.

**Comment by rmohr** [comment]
/reopen

We need to sync the content to an extra repo

**Comment by rmohr** [comment]
#2368 needs to be addressed too.

**Comment by cynepco3hahue** [comment]
@rmohr Will start to work on it, I believe we also need some documentation under the client-go repository and some examples.

**Comment by cynepco3hahue** [comment]
@fabiand Can you create a new repository for `client-go`?

**Comment by rmohr** [comment]
/close

fixed in #2510 and #2514.
